wrapper.py

```diff
@@ -0,0 +1,1278 @@
+        # XXX: map numpy dtypes
+        ar_tuple = tuple(ar.type(dtype) for ar in ar_typle)
+    return torch.cat(ar_tuple, axis, out=out)
+
+
+def squeeze(a, axis=None):
+    if axis is None:
+        return torch.squeeze(a)
+    else:
+        return torch.squeeze(a, axis)
+
+
+def bincount(x, /, weights=None, minlength=0):
+    return torch.bincount(x, weights, minlength)
+
+
+###### module-level queries of object properties
+
+def ndim(a):
+    return torch.as_tensor(a).ndim
+
+
+def shape(a):
+    return tuple(torch.as_tensor(a).shape)
+
+
+def size(a, axis=None):
+    if axis is None:
+        return torch.as_tensor(a).numel()
+    else:
+        return torch.as_tensor(a).shape[axis]
+
+
+###### shape manipulations and indexing
+
+def reshape(a, newshape, order='C'):
+    if order != 'C':
+        raise NotImplementedError
+    return torch.reshape(a, newshape)
+
+
+def broadcast_to(array, shape, subok=False):
+    if subok:
+        raise NotImplementedError
+    return torch.broadcast_to(array, shape)
+
+
+from torch import broadcast_shapes
+
+
+def broadcast_arrays(*args, subok=False):
+    if subok:
+        raise NotImplementedError
+    return torch.broadcast_tensors(*args)
+
+
+def unravel_index(indices, shape, order='C'):
+# cf https://github.com/pytorch/pytorch/pull/66687
+# this version is from 
+# https://discuss.pytorch.org/t/how-to-do-a-unravel-index-in-pytorch-just-like-in-numpy/12987/3
+    if order != 'C':
+        raise NotImplementedError
+    result = []
+    for index in indices:
+        out = []
+        for dim in reversed(shape):
+            out.append(index % dim)
+            index = index // dim
+        result.append(tuple(reversed(out)))
+    return result
+
+
+def ravel_multi_index(multi_index, dims, mode='raise', order='C'):
+    # XXX: not available in pytorch, implement
+    return sum(idx*dim for idx, dim in zip(multi_index, dims))
+
+
+
+###### reductions
+
+def argmax(a, axis=None, out=None, *, keepdims=NoValue):
+    if axis is None:
+        result = torch.argmax(a, keepdims=bool(keepdims))
+    else:
+        result = torch.argmax(a, axis, keepdims=bool(keepdims))
+    if out is not None:
+        out[...] = result
+    return result
+
+
+##### math functions
+
+from _unary import *
+abs = absolute
+
+
+def angle(z, deg=False):
+    result = torch.angle(z)
+    if deg:
+        result *= 180 / torch.pi
+    return result
+
+from torch import imag, real
+
+
+def real_if_close(a, tol=100):
+    if not torch.is_complex(a):
+        return a
+    if torch.abs(torch.imag) < tol * torch.finfo(a.dtype).eps:
+        return torch.real(a)
+    else:
+        return a
+
+
+def iscomplex(x):
+    if torch.is_complex(x):
+        return torch.as_tensor(x).imag == 0
+    result = torch.zeros_like(x, dtype=torch.bool)
+    return result[()]
+
+
+def isreal(x):
+    if torch.is_complex(x):
+        return torch.as_tensor(x).imag == 0
+    result = torch.zeros_like(x, dtype=torch.bool)
+    return result[()]
```

Comment:
What does this syntax mean? Why not simply return `result`?

Reply:
This is a "nice" numpy-ism for getting python scalars out of zero-dim arrays (or array scalar, I never know which one is which). Observe:

```
In [61]: a = np.array(3)

In [62]: a.shape
Out[62]: ()

In [63]: a[0]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [63], in <cell line: 1>()
----> 1 a[0]

IndexError: too many indices for array: array is 0-dimensional, but 1 were indexed

In [64]: a[()]
Out[64]: 3
```

This particular line is straight from `In [67]: print(inspect.getsource(np.iscomplex))`

Reply:
A right. Then you may want to do `a.item()` in PyTorch. This will work for any tensor with 1 element.

Reply:
Does it then need a separate check for one or more elements?
In numpy, it's a no-op
```
In [86]: np.arange(3)[()]
Out[86]: array([0, 1, 2])
```

Reply:
It does, yes. Note that this `a[()]` is pretty much the identity function (it does not return a scalar).

I guess this is another case where you may want to add a decorator `unpack_scalar` or so with this logic, as we never do it in PyTorch but it's quite a common operation in NumPy.

Reply:
Am going to keep this conversation unresolved so that this point does not fall under the radar.

However @rgommers would you know the numpy plans about these things? Is this what Sebastian plans to remove or will it stay?

Reply:
That will stay AFAIK. `unpack_scalar` to make this easier to understand seems like a good idea to me.